Created new allocation attributes for IBM storage quota

QuanMPhm opened the pull request:
Closes #231. More details in the commit message. 

This is a draft for now, since I have some details I wanted to confirm.

knikolla commented on src/coldfront_plugin_cloud/openshift.py:
```diff
@@ -74,6 +74,9 @@ class OpenShiftResourceAllocator(base.ResourceAllocator):
     QUOTA_KEY_MAPPING = {
         attributes.QUOTA_LIMITS_CPU: lambda x: {"limits.cpu": f"{x * 1000}m"},
         attributes.QUOTA_LIMITS_MEMORY: lambda x: {"limits.memory": f"{x}Mi"},
+        attributes.QUOTA_REQUESTS_IBM_STORAGE: lambda x: {
+            "requests.ibm.storage": f"{x}Gi"
+        },  # TODO: Confirm Openshift quota name for IBM storage
```
`<storage-class-name>.storageclass.storage.k8s.io/requests.storage`

https://docs.redhat.com/en/documentation/openshift_container_platform/4.19/html/building_applications/quotas#quotas-resources-managed_quotas-setting-per-project

where as of now the storage class name is `ibm-spectrum-scale-fileset`

knikolla replied:
`requests.storage` below will also need a migration to instead apply the existing value to  `ocs-external-storagecluster-ceph-rbd.storageclass.storage.k8s.io/requests.storage`. Validate allocations need to also be updated to migrate the quota.

knikolla replied:
> @knikolla What about `limits.ephemeral-storage`?

Leave it as it is.

knikolla commented on src/coldfront_plugin_cloud/attributes.py:
```diff
@@ -85,8 +85,13 @@ class CloudAllocationAttribute:
+)
+QUOTA_REQUESTS_STORAGE = "OpenShift Request on NESE Storage Quota (GiB)"
+QUOTA_REQUESTS_IBM_STORAGE = (
+    "OpenShift Request on IBM Storage Quota (GiB)"  # TODO: Acceptable names?
```
`OpenShift Request on IBM Spectrum Scale Storage Quota (GiB)

QuanMPhm commented:
@knikolla @joachimweyl @naved001 I have two main questions:
-  Because we are introducing a new allocation attribute, what should be the default behavior if a coldfront allocation does not already have the ibm storage quota attribute? Should it be set to 0 storage?
- For newly made allocations, what should be the static and multiplier amount for IBM storage. Basically, what goes [here](https://github.com/nerc-project/coldfront-plugin-cloud/blob/7e66279dc7a55bfe0832b4b1b740898dfe7db652/src/coldfront_plugin_cloud/tasks.py#L22-L71)?

naved001 commented on src/coldfront_plugin_cloud/attributes.py:
```diff
@@ -85,8 +87,13 @@ class CloudAllocationAttribute:
+QUOTA_LIMITS_EPHEMERAL_STORAGE_GB = (
+    "OpenShift Limit on NESE Ephemeral Storage Quota (GiB)"
+)
+QUOTA_REQUESTS_STORAGE = "OpenShift Request on NESE Storage Quota (GiB)"
```
I think you should call this QUOTA_REQUEST_NESE_STORAGE to make it clear that it's not the general storage quota.

naved001 commented on src/coldfront_plugin_cloud/attributes.py:
```diff
@@ -85,8 +87,13 @@ class CloudAllocationAttribute:
-QUOTA_LIMITS_EPHEMERAL_STORAGE_GB = "OpenShift Limit on Ephemeral Storage Quota (GiB)"
-QUOTA_REQUESTS_STORAGE = "OpenShift Request on Storage Quota (GiB)"
+QUOTA_LIMITS_EPHEMERAL_STORAGE_GB = (
+    "OpenShift Limit on NESE Ephemeral Storage Quota (GiB)"
```
I don't think the ephemeral storage comes out of NESE, it's the local storage which is why it's ephemeral.

naved001 reviewed (commented):
`requests.storage` will that be migrated and become `ocs-external-storagecluster-ceph-rbd.storageclass.storage.k8s.io/requests.storage` for all the old allocations?

naved001 commented on src/coldfront_plugin_cloud/attributes.py:
```diff
@@ -24,6 +24,7 @@ class CloudAllocationAttribute:
 RESOURCE_API_URL = "OpenShift API Endpoint URL"
 RESOURCE_IDENTITY_NAME = "OpenShift Identity Provider Name"
 RESOURCE_ROLE = "Role for User in Project"
+RESOURCE_IBM_AVAILABLE = "IBM Spectrum Scale Storage Available"
```
> or whether or not users can choose to not use virtualization in a cluster that supports virtualization?

If an OpenShift cluster is configured to support virtualization then a user can still choose to not run VMs and only run pods. 

Also, right now we are not offering virtualization at all because we are waiting for some features to land, but if we did there will be a dedicated cluster for openshift virtualization.

QuanMPhm commented:
> `requests.storage` will that be migrated and become `ocs-external-storagecluster-ceph-rbd.storageclass.storage.k8s.io/requests.storage` for all the old allocations?

@naved001 Sorry for being late to answer your question. Yes. I've updated `register_cloud_attributes.py` so that running the command will migrate the NESE storage attribute. As for the IBM storage attribute, once we enable IBM storage on the `NERC-OCP` resource, running `validate_alloacations.py` should automatically add the IBM attribute to all pre-existing `NERC-OCP` allocations.

I've addressed all your comments so far.

knikolla commented on src/coldfront_plugin_cloud/tasks.py:
```diff
@@ -74,7 +76,17 @@
+    resource_expected_quotas = UNIT_QUOTA_MULTIPLIERS[resource_name].copy()
+
+    is_ibm_storage_available = (
+        allocator.resource.get_attribute(attributes.RESOURCE_IBM_AVAILABLE) == "true"
```
Please make this case insensitive for the comparison.

knikolla reviewed (changes requested):
One comment, otherwise looks good! Great job!

QuanMPhm commented:
@knikolla The comparison is now case-insensitive

knikolla commented:
@QuanMPhm fails testing on `allocator.resource.get_attribute(attributes.RESOURCE_IBM_AVAILABLE).lower()` not working for `None` values.
